app/controllers/application_controller.rb

```diff
@@ -1,3 +1,29 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery
+
+  before_action :require_login
+  helper_method :logged_in?, :current_user
+
+  def set_current_user(user = nil)
+    if user
+      session[:user_id] = user.id
+    else
+      session[:user_id] = user
+    end
+  end
+
+  def current_user
+    User.find(session[:user_id]) if session[:user_id]
```

Comment:
This method is not finished - you should memoize current_user to an instance variable.

`@current_user ||= User.find(session[:user_id]) if session[:user_id]`